group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -5524,5 +5670,2897 @@ private static Record newGroupMetadataRecord(
+            context,
+            "group-id",
+            "leader-instance-id",
+            "follower-instance-id"
+        );
+
+        // A third member joins. Trigger a rebalance.
+        JoinGroupRequestData request = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withDefaultProtocolTypeAndProtocols()
+            .build();
+
+        context.sendGenericGroupJoin(request, new CompletableFuture<>());
+
+        assertTrue(group.isInState(PREPARING_REBALANCE));
+
+        // Old follower rejoins group will be matching current member.id.
+        CompletableFuture<JoinGroupResponseData> oldFollowerJoinFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(
+            request
+                .setMemberId(rebalanceResult.followerId)
+                .setGroupInstanceId("follower-instance-id"),
+            oldFollowerJoinFuture);
+
+        assertTrue(result.records().isEmpty());
+        assertFalse(oldFollowerJoinFuture.isDone());
+
+        // Duplicate follower joins group with unknown member id will trigger member id replacement.
+        result = context.sendGenericGroupJoin(
+            request.setMemberId(UNKNOWN_MEMBER_ID).setGroupInstanceId("follower-instance-id"),
+            new CompletableFuture<>());
+
+        // Old member shall be fenced immediately upon duplicate follower joins.
+        assertTrue(result.records().isEmpty());
+        assertTrue(oldFollowerJoinFuture.isDone());
+
+        JoinGroupResponseData expectedResponse = new JoinGroupResponseData()
+            .setErrorCode(Errors.FENCED_INSTANCE_ID.code())
+            .setProtocolName(null)
+            .setProtocolType(null)
+            .setLeader(UNKNOWN_MEMBER_ID)
+            .setMemberId(rebalanceResult.followerId)
+            .setGenerationId(-1);
+
+        checkJoinGroupResponse(
+            expectedResponse,
+            oldFollowerJoinFuture.get(),
+            group,
+            PREPARING_REBALANCE,
+            Collections.emptySet()
+        );
+    }
+
+    @Test
+    public void testStaticMemberFenceDuplicateSyncingFollowerAfterMemberIdChanged() throws Exception {
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .build();
+        GenericGroup group = context.createGenericGroup("group-id");
+
+        RebalanceResult rebalanceResult = staticMembersJoinAndRebalance(
+            context,
+            "group-id",
+            "leader-instance-id",
+            "follower-instance-id"
+        );
+
+        // Known leader rejoins will trigger rebalance.
+        JoinGroupRequestData request = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withGroupInstanceId("leader-instance-id")
+            .withMemberId(rebalanceResult.leaderId)
+            .withDefaultProtocolTypeAndProtocols()
```

Comment:
It would be if it was the follower. i updated to `withProtocolSuperset()` so to be precise.

A join group from the leader during stable state triggers a rebalance to update assignment

```
                        // Force a rebalance if the leader sends JoinGroup;
                        // This allows the leader to trigger rebalances for changes affecting assignment
                        // which do not affect the member metadata (such as topic metadata changes for the consumer)
```

Reply:
Ah right. That makes sense.